### PR TITLE
Migration/stavenote2

### PR DESCRIPTION
**Pull request by rvilarl**
Migration of StaveNote, it was not easy with those combination of boolean|number. The reference test has helped a lot!
I have set the wiki in my todo list. But this should definitvely be a health check before merging.

### CODE REVIEW

File: src/stavenote.ts [hunk 1]
```diff
@@ -183,7 +244,7 @@ export class StaveNote extends StemmableNote {
     //
     // Special case 1 :: middle voice rest between two notes
     //
-    if (noteM.isrest && !noteU.isrest && !noteL.isrest) {
+    if (noteM?.isrest && !noteU.isrest && !noteL.isrest) {
```
Comment (0xfe) [inline src/stavenote.ts:247]: `noteM` must be defined here, so better to add a `if (!noteM) throw...` and remove the `noteM?`

File: src/stavenote.ts [hunk 2]
```diff
@@ -274,13 +336,14 @@ export class StaveNote extends StemmableNote {
       //
       // We also extend the y for each note by a half notehead because the
       // notehead's origin is centered
-      const topNoteBottomY = topNote.getStave().getYForLine(5 - topKeys[0].line + HALF_NOTEHEAD_HEIGHT);
+      const topNoteBottomY = topNote.getStave()?.getYForLine(5 - topKeys[0].line + HALF_NOTEHEAD_HEIGHT);
```
Comment (0xfe) [inline src/stavenote.ts:339]: `topNote.getStave()` must be defined, so add an exception check here and remove the `?`.

File: src/stavenote.ts [hunk 3]
```diff
@@ -274,13 +336,14 @@ export class StaveNote extends StemmableNote {
...
 
       const bottomNoteTopY = bottomNote
         .getStave()
```
Comment (0xfe) [inline src/stavenote.ts:342]: Same here:

```typescript
const bottomStave = bottomNote.getStave();
if (!bottomStave) throw "NoStave"...
```

File: src/stavenote.ts [hunk 4]
```diff
@@ -925,86 +1005,87 @@ export class StaveNote extends StemmableNote {
...
         highestNonDisplacedLine = Math.max(line, highestNonDisplacedLine);
         lowestNonDisplacedLine = Math.min(line, lowestNonDisplacedLine);
       }
     }, this);
 
     return {
-      y_top: yTop,
-      y_bottom: yBottom,
+      y_top: yTop ?? 0,
+      y_bottom: yBottom ?? 0,
```
Comment (0xfe) [inline src/stavenote.ts:1057]: Do you needs these `?? 0` here? Both `yTop` and `yBottom` will be defined if we get here.

File: src/types/common.d.ts [hunk 5]
```diff
@@ -98,11 +98,19 @@ export interface RenderContext {
...
+  openGroup(...args: any[]):any;
   closeGroup(): void;
 
   /**
    * canvas returns TextMetrics, SVG returns SVGRect, Raphael returns {width : number, height : number}. Only width is used throughout VexFlow.
    */
   measureText(text: string): { width: number };
 }
+
+export interface ModifierContextState {
```
Comment (0xfe) [inline src/types/common.d.ts:111]: Let's move this to `modifiercontext.ts`?

**Review by 0xfe — COMMENTED** [review]
Looks good overall. Last few comments I hope :-)

File: src/stavenote.ts (hunk 1 above)
Comment (rvilarl) [reply src/stavenote.ts:247]: done

File: src/stavenote.ts (hunk 2 above)
Comment (rvilarl) [reply src/stavenote.ts:339]: done

File: src/stavenote.ts (hunk 3 above)
Comment (rvilarl) [reply src/stavenote.ts:342]: done

File: src/stavenote.ts (hunk 4 above)
Comment (rvilarl) [reply src/stavenote.ts:1057]: done by initialization to numbers beyond limits 100000 & -100000.

File: src/types/common.d.ts (hunk 5 above)
Comment (rvilarl) [reply src/types/common.d.ts:111]: Comment added

**Comment by rvilarl** [comment]
Ready

File: src/stavenote.ts [hunk 6]
```diff
@@ -925,46 +1008,47 @@ export class StaveNote extends StemmableNote {
...
-  getNoteHeadBounds() {
+  getNoteHeadBounds(): StaveNoteHeadBounds {
+    if (!this.stave) throw new Vex.RERR('NoStave', 'No stave attached to this note.');
     // Top and bottom Y values for stem.
-    let yTop = null;
-    let yBottom = null;
-    let nonDisplacedX = null;
-    let displacedX = null;
+    let yTop: number = 100000;
+    let yBottom: number = -100000;
```
Comment (0xfe) [inline src/stavenote.ts:1015]: Better to use `+Infinity` and `-Infinity` here.

File: src/stavenote.ts [hunk 7]
```diff
@@ -925,46 +1008,47 @@ export class StaveNote extends StemmableNote {
...
     let highestNonDisplacedLine = highestLine;
     let lowestNonDisplacedLine = lowestLine;
 
     this.note_heads.forEach((notehead) => {
-      const line = notehead.getLine();
+      const line: number = notehead.getLine();
       const y = notehead.getY();
 
-      if (yTop === null || y < yTop) {
+      if (y < yTop) {
```
Comment (0xfe) [inline src/stavenote.ts:1030]: `yTop = Math.min(y, yTop)`

File: src/stavenote.ts [hunk 8]
```diff
@@ -925,46 +1008,47 @@ export class StaveNote extends StemmableNote {
...
         yTop = y;
       }
 
-      if (yBottom === null || y > yBottom) {
+      if (y > yBottom) {
```
Comment (0xfe) [inline src/stavenote.ts:1034]: `yBottom = Math.max(y, yBottom)`

File: src/stavenote.ts [hunk 9]
```diff
@@ -978,33 +1062,33 @@ export class StaveNote extends StemmableNote {
       non_displaced_x: nonDisplacedX,
       highest_line: highestLine,
       lowest_line: lowestLine,
-      highest_displaced_line: highestDisplacedLine,
-      lowest_displaced_line: lowestDisplacedLine,
+      highest_displaced_line: highestDisplacedLine ?? 0,
```
Comment (0xfe) [inline src/stavenote.ts:1065]: I think these are wrong -- if there's no displaced lines, then returning `0` would be incorrect -- better to return `undefined`.

File: src/stavenote.ts (hunk 6 above)
Comment (rvilarl) [reply src/stavenote.ts]: Ok

File: src/stavenote.ts (hunk 7 above)
Comment (rvilarl) [reply src/stavenote.ts]: Ok

**Comment by 0xfe** [comment]
Great!

**Pull request merged**